### PR TITLE
Service graph processor

### PR DESCRIPTION
**Pull request by mapno**
**Description:** First draft of the service graph processor.

The service graphs processor is a traces processor that builds a map representing the interrelationships between various services in a system.
The processor will analyse trace data and generate metrics describing the relationship between the services.
These metrics can be used by data visualization apps (e.g. Grafana) to draw a service graph.

Service graphs are useful for a number of use-cases:

* Infer the topology of a distributed system. As distributed systems grow, they become more complex. Service graphs can help you understand the structure of the system.
* Provide a high level overview of the health of your system.
Service graphs show error rates, latencies, among other relevant data.
* Provide an historic view of a system’s topology.
Distributed systems change very frequently,
and service graphs offer a way of seeing how these systems have evolved over time.

This component is based on [Grafana Tempo's service graph processor](https://github.com/grafana/tempo/tree/main/modules/generator/processor/servicegraphs).

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9232

**Testing:** Unit tests are added across all relevant files (mainly `processor_test.go`).

### CODE REVIEW

**Comment by JaredTan95** [comment]
I think `make goporto` may be helpful to fix CI/checks

**Comment by JaredTan95** [comment]
> I think `make goporto` may be helpful to fix CI/checks

and `make lint` to fix linter :-P

**Comment by jpkrohling** [comment]
@mapno, do you need any help moving forward with this ?

**Comment by sarahsporck** [comment]
Any updates?

**Comment by mapno** [comment]
Hey, sorry, I've been busy with other work and struggled to find time for this. Not a lot is missing, mainly documentation and updating it to the latest changes we've made on Tempo. I'll try to do it in the coming days. Apologies 🙏

**Comment by nicolastakashi** [comment]
Hey @mapno
I have a doubt.
Will this implementation solve the current issue of creating service graphs for Zipkin traces that we've on the Grafana Tempo metrics generator? https://github.com/grafana/tempo/issues/1495#issuecomment-1208075036

**Comment by JaredTan95** [comment]
> Hey @mapno I have a doubt. Will this implementation solve the current issue of creating service graphs for Zipkin traces that we've on the Grafana Tempo metrics generator? [grafana/tempo#1495](https://github.com/grafana/tempo/issues/1495)

If using otel col receive Zipkin traces, this processor will graph metrics and Grafana Tempo cloud use those metrics to create graphs.

**Comment by mapno** [comment]
@jpkrohling updated the branch with some changes and documentation, could it be opened again?

File: processor/servicegraphprocessor/processor.go [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+	nextConsumer    consumer.Traces
+	metricsExporter consumer.Metrics
+
+	store store.Store
+
+	startTime time.Time
+
+	seriesMutex                    sync.Mutex
+	reqTotal                       map[string]int64
+	reqFailedTotal                 map[string]int64
```
Comment (JaredTan95) [inline processor/servicegraphprocessor/processor.go:72]: Looks like `reqFailedTotal` is not being used, should remove it?

**Pull request closed without merging**